src/geom/SipApproximation.cc

```diff
@@ -0,0 +1,340 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <algorithm>
+#include <vector>
+
+#include "Eigen/SVD"
+#include "Eigen/QR"
+#include "lsst/afw/geom/SipApproximation.h"
+#include "lsst/geom/polynomials/PolynomialFunction2d.h"
+
+namespace lsst { namespace afw { namespace geom {
+
+namespace poly = lsst::geom::polynomials;
+
+namespace {
+
+std::pair<poly::PolynomialFunction2dYX, poly::PolynomialFunction2dYX> fitSipOneDirection(
+    int order,
+    Box2D const & box,
+    double svdThreshold,
+    std::vector<Point2D> const & input,
+    std::vector<Point2D> const & output
+) {
+    // The scaled polynomial basis evaluates polynomials after mapping the
+    // input coordinates from the given box to [-1, 1]x[-1, 1] (for numerical
+    // stability).
+    auto basis = poly::ScaledPolynomialBasis2dYX(order, box);
+    auto workspace = basis.makeWorkspace();
+    Eigen::MatrixXd matrix = Eigen::MatrixXd::Zero(input.size(), basis.size());
+    Eigen::VectorXd xRhs(input.size());
+    Eigen::VectorXd yRhs(input.size());
+    for (int i = 0; i < matrix.rows(); ++i) {
+        basis.fill(input[i], matrix.row(i), workspace);
+        auto rhs = output[i] - input[i];
+        xRhs[i] = rhs.getX();
+        yRhs[i] = rhs.getY();
+    }
+    // Since we're not trying to null the zeroth- and first-order terms, the
+    // solution is just linear least squares, and we can do that with SVD.
+    auto decomp = matrix.jacobiSvd(Eigen::ComputeThinU | Eigen::ComputeThinV);
+    if (svdThreshold >= 0) {
+        decomp.setThreshold(svdThreshold);
+    }
+    auto scaledX = makeFunction2d(basis, decomp.solve(xRhs));
+    auto scaledY = makeFunction2d(basis, decomp.solve(yRhs));
+    // On return, we simplify the polynomials by moving the remapping transform
+    // into the coefficients themselves.
+    return std::make_pair(unscaled(scaledX), unscaled(scaledY));
+}
+
+// Return a vector of points on a grid, covering the given bounding box.
+std::vector<Point2D> makeGrid(Box2D const & bbox, Extent2I const & dimensions) {
+    std::vector<Point2D> points;
+    points.reserve(dimensions.getX()*dimensions.getY());
+    double const dx = bbox.getWidth()/dimensions.getX();
+    double const dy = bbox.getHeight()/dimensions.getY();
+    for (int iy = 0; iy < dimensions.getY(); ++iy) {
+        double const y = bbox.getMinY() + iy*dy;
+        for (int ix = 0; ix < dimensions.getX(); ++ix) {
+            points.emplace_back(bbox.getMinX() + ix*dx, y);
+        }
+    }
+    return points;
+}
+
+poly::PolynomialFunction2dYX readFunctionFromCoeffMatrix(ndarray::Array<double const, 2> const & coeffs) {
```

Comment:
Please add brief documentation of what this does. Why this name instead of, say, "makePolynomial"?